### PR TITLE
fix: faster response to config map changes

### LINKED ISSUE #16: Values are not updated on module discover error loop
Module discovery can get stuck in error loop, i.e. on enabled script crush. In this case values for enabled script is not updated between enabled script executions.
This happens on startup.

### LINKED ISSUE #43: Allow disabling of module having errors (during it's installation or in hooks)
The current implementation has possible deadlock: if the module has errors, either in the helm templates or in hooks, the tasks are stuck in the queue and disabling the module in ConfigMap don't help, because of module tasks are not removed from the queue.

Proposed change: When the module becomes disabled (in ConfigMap, or by any other means) the following should happen (in the following order):
1. All informers should be stopped and all crontab bindings should be disabled, so no new tasks (possibly containing errors) will be added to the queue.
2. All existing module tasks should be removed from the queue.
3. Tasks associated with module disabling should be added (helm uninstall, and afterDeleteHelm hooks)

### LINKED ISSUE #184: Values file renders after enabled-script execution
**Expected behavior (what you expected to happen)**:
Enabled-script has actual values during execution.

**Actual behavior (what actually happened)**:
Enabled-script has values lagging by one iteration. Seems like values are rendered after script execution

**Steps to reproduce**:
[Here](https://github.com/yalosev/aop/tree/enabled-script) you can find all source code and steps for reproducing

**Environment**:
- Addon-operator version: flant/addon-operator:alpine3.12
- Helm version: - 
- Kubernetes version: 1.19.7
- Installation type (kubectl apply, helm chart, etc.): kubectl apply

### LINKED ISSUE #233: Run ModulePurge tasks before ModuleRun
**Expected behavior (what you expected to happen)**:

The disabled or absent modules are purged before running enabled modules.

**Actual behavior (what actually happened)**:

The absent module is purged after enabled modules run. It prevents module renaming and Kubernetes resourses adoption.

**Steps to reproduce**:
1. ...
2. ...
3. ...

**Environment**:
- Addon-operator version: 1.0.0.-rc.1
- Helm version: 3
- Kubernetes version: 1.19
- Installation type (kubectl apply, helm chart, etc.): deckhouse

**Anything else we should know?**:

**Additional information for debugging (if necessary)**:

<details><summary>Hook script</summary><br><pre>

</pre></details>

<details><summary>Logs</summary><br><pre>

</pre></details>

### PR DESCRIPTION
**Pull request by diafour**
#### Overview

Huge refactoring to speedup reactions to ConfigMap changes. Fixes in enabled state
calculations.

KubeConfigManager is now responsible to signal if config is changed or invalid.

- fix: Update metric if ConfigMap is invalid (malformed module name or module values YAML)

ModuleManager is now the only component that runs enabled scripts. It is responsible
for calculating ModulesReload/ReloadAll event. It lists Helm releases once at start
to detect initial enabled modules and modules to purge.

- fix: Fresh config values for enabled scripts (#184, #16)

AddonOperator uses task ConvergeModules to handle changes in ConfigMap or global values
and reload all modules or only changed.

- fix: ModulePurge tasks moved between GlobalSynchronization and first ConvergeModules (#233)
- fix: Clear queues from tasks for disabled module or errored hooks on ConfigMap changes (#43)
- ref: ConvergeModules is a new multi-phase task.

Other:
- mocks for HelmResourcesManager and KubeConfigManager
- Helm struct instead of global variables helm.NewClient and helm.HealthzHandler

#### What this PR does / why we need it

More responsive reactions to ConfigMap changes and more clear "enabled state" calculations.

Closes #184, closes #16, closes #233, closes #43.

#### TODO

- [x] add queue cleaning after ConvergeModules task
- [ ] check documentation for deprecated terms (new task is introduced: ConvergeModules)
- [ ] Update shell-operator in go.mod after merging https://github.com/flant/shell-operator/pull/388

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/addon-operator/metrics.go [hunk 1]
```diff
@@ -133,3 +136,26 @@ func RegisterHookMetrics(metricStorage *metric_storage.MetricStorage) {
 			"queue":   "",
 		})
 }
+
+func StartLiveTicksUpdater(metricStorage *metric_storage.MetricStorage) {
+	// Addon-operator live ticks.
+	go func() {
+		for {
+			metricStorage.CounterAdd("{PREFIX}live_ticks", 1.0, map[string]string{})
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/addon-operator/metrics.go:144]: `CounterAdd` belongs to the first family of methods to work with internal metrics. These methods are thin wrappers around methods from prometheus-client and they are thread safe. MetricsCollector is a later invention for metrics from hooks.

File: pkg/addon-operator/http_server.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		if helmHealthHandler == nil {
+			writer.WriteHeader(http.StatusOK)
+			return
+		}
+		helmHealthHandler(writer, request)
+	})
+
+	http.HandleFunc("/ready", func(w http.ResponseWriter, request *http.Request) {
+		if op.IsStartupConvergeDone() {
+			w.WriteHeader(200)
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/addon-operator/http_server.go:38]: Done

File: pkg/addon-operator/http_server.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			_, _ = w.Write([]byte("Startup converge done.\n"))
+		} else {
+			w.WriteHeader(500)
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/addon-operator/http_server.go:41]: Done

File: pkg/kube_config_manager/checksums.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package kube_config_manager
+
+// Checksums is a non-safe storage if checksums for KubeConfig sections.
```
[the remark this answers is not in the document]
Comment (diafour) [reply pkg/kube_config_manager/checksums.go]: Fixed!

**Review by yalosev — APPROVED** [review]

**Pull request merged**